### PR TITLE
GRF-220 : Upsert Category command handler

### PR DESCRIPTION
**Pull request by antoinetrouve**
**Description (for Contributor and Core Developer)**

- Create the upsert command handler for the update.
- /!\ The Creation process still to do and the saver (see [GRF-221](https://github.com/akeneo/pim-community-dev/pull/17617)) need to be include when ready.

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Category/back/Application/UpsertCategoryCommandHandler.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function __invoke(UpsertCategoryCommand $command): void
+    {
+        $this->validateCommand($command);
+
+        $category = $this->findCategoryByCode->__invoke($command->categoryCode());
```
Comment (BryanS29-akeneo) [inline src/Akeneo/Category/back/Application/UpsertCategoryCommandHandler.php:43]: we don't have to call exactly `__invoke`,  you can replace it by `$category = ($this->findCategoryByCode)($command->categoryCode());`

**Review by BryanS29-akeneo — APPROVED** [review]

File: src/Akeneo/Category/back/Application/UpsertCategoryCommandHandler.php (hunk 1 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/Application/UpsertCategoryCommandHandler.php:43]: done

File: src/Akeneo/Category/back/Application/UpsertCategoryCommandHandler.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        $category = ($this->findCategoryByCode)($command->categoryCode());
+
+        $isCreation = null === $category;
+        if ($isCreation) {
+            throw new Exception("Command to create a category is in progress.");
+        }
+
+        $this->updateCategory($category, $command);
+
+        $this->updater->update($category, $command->userIntents());
```
Comment (olivierpontier) [inline src/Akeneo/Category/back/Application/UpsertCategoryCommandHandler.php:52]: The naming is confusing, by reading the code I don't understand the difference between `updateCategory` and calling the updater. Is the second service more like a saver in charge of storing the updated category object in the database ?

File: src/Akeneo/Category/back/tests/Specification/Infrastructure/Validation/LocalizeUserIntentsShouldBeUniqueValidatorSpec.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        $context->buildViolation(Argument::cetera())->shouldNotBeCalled();
+
+        $this->validate([
+            new SetLabel('fr_FR', 'libelle'),
+            new SetLabel('en_US', 'label'),
+            new SetCode('code')
+        ], new LocalizeUserIntentsShouldBeUnique());
+    }
+
+    function it_throw_an_exception_when_the_localize_intents_are_not_distinct(ExecutionContext $context, ConstraintViolationBuilderInterface $violationBuilder)
```
Comment (olivierpontier) [inline src/Akeneo/Category/back/tests/Specification/Infrastructure/Validation/LocalizeUserIntentsShouldBeUniqueValidatorSpec.php:51]:
```suggestion
    function it_throws_an_exception_when_the_localize_intents_are_not_distinct(ExecutionContext $context, ConstraintViolationBuilderInterface $violationBuilder)
```

File: src/Akeneo/Category/back/Api/Command/UserIntents/LocalizeUserIntent.php [hunk 4]
```diff
@@ -2,15 +2,13 @@
 
 declare(strict_types=1);
 
-namespace Akeneo\Category\Api\Command;
+namespace Akeneo\Category\Api\Command\UserIntents;
```
Comment (olivierpontier) [inline src/Akeneo/Category/back/Api/Command/UserIntents/LocalizeUserIntent.php:5]: Weird, it seems the DummyHandler class has been renamed into LocalizeUserIntent. It is due to a conflict resolution?

Comment (antoinetrouve) [reply src/Akeneo/Category/back/Api/Command/UserIntents/LocalizeUserIntent.php:5]: No. The localizeUserIntent is an interface for the userIntent. Not directly related to the dummy handler.

File: src/Akeneo/Category/back/Application/UpsertCategoryCommandHandler.php (hunk 2 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/Application/UpsertCategoryCommandHandler.php:52]: I just renaming to saver which is more corresponding to its usage.

Comment (antoinetrouve) [reply src/Akeneo/Category/back/Application/UpsertCategoryCommandHandler.php:52]:
```suggestion
        $this->updateCategory($category, $command);

        $this->saver->update($category, $command->userIntents());
```

File: src/Akeneo/Category/back/Api/Command/UserIntents/LocalizeUserIntent.php (hunk 4 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/Api/Command/UserIntents/LocalizeUserIntent.php:5]: This interface is used to identify the user intent localizable. For now, It's mainly used to validate the userIntents given in parameter in the command. see validator.yml confg

Comment (antoinetrouve) [reply src/Akeneo/Category/back/Api/Command/UserIntents/LocalizeUserIntent.php:5]: done

File: src/Akeneo/Category/back/tests/Specification/Infrastructure/Validation/LocalizeUserIntentsShouldBeUniqueValidatorSpec.php (hunk 3 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/tests/Specification/Infrastructure/Validation/LocalizeUserIntentsShouldBeUniqueValidatorSpec.php:51]: updated

File: src/Akeneo/Category/back/Infrastructure/Validation/LocalizeUserIntentsShouldBeUniqueValidator.php [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+ */
+final class LocalizeUserIntentsShouldBeUniqueValidator extends ConstraintValidator
+{
+
+    /**
+     * @param UserIntent[] $value
+     * @param Constraint $constraint
+     * @return void
+     */
+    public function validate($value, Constraint $constraint)
```
Comment (etiennekerbastard) [inline src/Akeneo/Category/back/Infrastructure/Validation/LocalizeUserIntentsShouldBeUniqueValidator.php:26]: public function validate(**array** $value, Constraint $constraint): **void**

File: src/Akeneo/Category/back/tests/Specification/Application/UpsertCategoryCommandHandlerSpec.php [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        $validator->validate($command)->shouldBeCalledOnce()->willReturn(new ConstraintViolationList());
+
+        $findCategoryByCode->__invoke('code')->shouldBeCalledOnce()->willReturn($category);
+        $saver->update($category, $command->userIntents())->shouldBeCalledOnce();
+        $eventDispatcher->dispatch($event)->shouldBeCalledOnce()->willReturn($event);
+
+        $this->__invoke($command);
+    }
+
+    function it_creates_and_save_a_category(
```
Comment (etiennekerbastard) [inline src/Akeneo/Category/back/tests/Specification/Application/UpsertCategoryCommandHandlerSpec.php:81]: function it_creates_and_save**s**_a_category(

**Review by etiennekerbastard — CHANGES_REQUESTED** [review]

File: src/Akeneo/Category/back/tests/Specification/Application/UpsertCategoryCommandHandlerSpec.php (hunk 6 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/tests/Specification/Application/UpsertCategoryCommandHandlerSpec.php:81]: done :)

File: src/Akeneo/Category/back/Infrastructure/Validation/LocalizeUserIntentsShouldBeUniqueValidator.php (hunk 5 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/Infrastructure/Validation/LocalizeUserIntentsShouldBeUniqueValidator.php:26]: I cannot change the function signature. I have to respect the original one. The declaration must be compatible with ConstraintValidatorInterface.

**Review by etiennekerbastard — APPROVED** [review]

**Pull request merged**